File: euler49.py
# Какое 12-значное число образуется, если объединить три члена
# этой прогрессии?

from datetime import datetime
from itertools import permutations
from typing import List, Optional


def eratosfen(number: int) -> List[int]:
    """
    Решето Эратосфена: генерация простых чисел вплоть до заданного числа.

    Args:
        number: Верхняя граница диапазона.

    Returns:
        Список простых чисел.
    """
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve


def twelve_digit(pan_primes_check: List[int]) -> Optional[str]:
    """
    Проверяет последовательность простых чисел на арифметическую прогрессию.

    Args:
        pan_primes_check: Список простых чисел для проверки.

    Returns:
        12-значное число из трех членов прогрессии или False.
    """
    for i in range(len(pan_primes_check)):
        for j in range(i + 1, len(pan_primes_check)):
            delta = pan_primes_check[j] - pan_primes_check[i]
            if pan_primes_check[j] + delta in pan_primes_check:
                return (
                    str(pan_primes_check[i])
                    + str(pan_primes_check[j])
                    + str(pan_primes_check[j] + delta)
                )
    return None


def euler49() -> None:
    """
    Решение задачи Эйлера №49.

    Какое 12-значное число образуется, если объединить три члена прогрессии?
    """
    n = 4

    # 1 вариант
    start_time = datetime.now()
    prime_permutation = []
    primes = [i for i in set(eratosfen(10 ** n)) if i > 999]
    for sets_primes in primes:
        pr_check = sorted(list(set([
            x for x in [int(''.join(x)) for x in permutations(str(sets_primes))]
            if x in primes
        ])))
        if len(pr_check) >= 3:
            found = twelve_digit(pr_check)
            if found and int(found) not in prime_permutation:
                prime_permutation.append(int(found))
    print(prime_permutation[-1])
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    pass
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler49()

File: go/euler49.go
// euler49 находит 12-значное число из трех членов арифметической
// прогрессии простых чисел.
//
// Решение задачи Эйлера №49.
package main

import (
	"fmt"
	"math"
	"sort"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func isPrime(number int) bool {
	if number < 2 {
		return false
	}
	if number == 2 || number == 3 || number == 5 || number == 7 {
		return true
	}
	if number%2 == 0 || number%3 == 0 || number%5 == 0 {
		return false
	}
	sqrt := int(math.Sqrt(float64(number)))
	for i := 5; i <= sqrt; i += 6 {
		if number%i == 0 || number%(i+2) == 0 {
			return false
		}
	}
	return true
}

func permutations(arr []int) [][]int {
	result := [][]int{{}}
	for _, elem := range arr {
		newResult := [][]int{}
		for _, permutation := range result {
			for i := 0; i <= len(permutation); i++ {
				newPermutation := make([]int, len(permutation)+1)
				copy(newPermutation, permutation[:i])
				newPermutation[i] = elem
				copy(newPermutation[i+1:], permutation[i:])
				newResult = append(newResult, newPermutation)
			}
		}
		result = newResult
	}
	return result
}

func twelveDigit(panPrimesCheck []int) string {
	for i := 0; i < len(panPrimesCheck); i++ {
		for j := i + 1; j < len(panPrimesCheck); j++ {
			delta := panPrimesCheck[j] - panPrimesCheck[i]
			if contains(panPrimesCheck, panPrimesCheck[j]+delta) {
				return fmt.Sprintf("%d%d%d",
					panPrimesCheck[i],
					panPrimesCheck[j],
					panPrimesCheck[j]+delta)
			}
		}
	}
	return ""
}

func euler49() {
	n := 4

	startTime := time.Now()
	sieve := eratosfen(powInt(10, n))
	primes := make([]int, 0)
	for i := 1000; i <= powInt(10, n); i++ {
		if sieve[i] {
			primes = append(primes, i)
		}
	}
	primePermutation := make([]int, 0)
	for _, setsPrimes := range primes {
		perms := permutations(digitsToSlice(setsPrimes))
		prCheck := make([]int, 0)
		seen := make(map[int]bool)
		for _, p := range perms {
			num := sliceToInt(p)
			if sieve[num] && !seen[num] {
				prCheck = append(prCheck, num)
				seen[num] = true
			}
		}
		sort.Ints(prCheck)
		if len(prCheck) >= 3 {
			found := twelveDigit(prCheck)
			if found != "" && !contains(primePermutation, toInt(found)) {
				primePermutation = append(primePermutation, toInt(found))
			}
		}
	}
	maxVal := 0
	for _, v := range primePermutation {
		if v > maxVal {
			maxVal = v
		}
	}
	fmt.Println(maxVal)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func contains(slice []int, val int) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func digitsToSlice(n int) []int {
	var result []int
	for n > 0 {
		result = append([]int{n % 10}, result...)
		n /= 10
	}
	return result
}

func sliceToInt(slice []int) int {
	result := 0
	for _, v := range slice {
		result = result*10 + v
	}
	return result
}

func toInt(s string) int {
	result := 0
	for _, c := range s {
		result = result*10 + int(c-'0')
	}
	return result
}

func powInt(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

func main() {
	euler49()
}
